### ISSUE TITLE
MOxUnit doesn't detect or runs tests

### ISSUE THREAD

**Issue by homocomputeris** [MOxUnit/MOxUnit#36]
Hi,
I've encountered an issue: MOxUnit doesn't see test suites.

```
>> version
ans =9.1.0.441655 (R2016b)
>> pwd
ans =/home/user/Documents/MATLAB/Add-Ons/MOxUnit/tests
>> moxunit_runtests -verbose
suite: 0 tests
--------------------------------------------------
OK
ans =  logical   1
```
```

>> path
/home/user/Documents/MATLAB/Add-Ons/MOxUnit
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/branches
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/hooks
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/info
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/logs
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/logs/refs
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/logs/refs/heads
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/logs/refs/remotes
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/logs/refs/remotes/origin
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/objects
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/objects/info
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/objects/pack
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/refs
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/refs/heads
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/refs/remotes
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/refs/remotes/origin
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/.git/refs/tags
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/MOxUnit
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/MOxUnit/util
/home/user/Documents/MATLAB/Add-Ons/MOxUnit/tests
```

**Comment by homocomputeris** [comment]
Moreover, in `test_test_suite.m`, for example, happens this:
```
>> initTestSuite
Index exceeds matrix dimensions.
Error in initTestSuite (line 44)
    caller_fn=call_stack(idx+1).file;
```

**Comment by nno** [comment]
Thanks for the report. I am using R2015a and with that version it runs all the tests perfectly.
Note that your path contains more elements than necessary; you probably did something like 
    ```
    addpath(genpath('/home/user/Documents/MATLAB/Add-Ons/MOxUnit')
    ```

instead of 

    ```
    cd('/home/user/Documents/MATLAB/Add-Ons/MOxUnit/MOxUnit');
    moxunit_set_path
    ```

but also when I use the ``genpath`` approach it runs all the tests. Do you have access to earlier version of Matlab (e.g. 2015a) to see if this issue is dependent on the version?

**Comment by nno** [comment]
Re the error when calling ``initTestSuite`` directly (not from inside a function), #38 should provide a more informative error message.

**Comment by homocomputeris** [comment]
Adding path as you propose doesn't change anything, unfortunately.

I had never used moxunit with Matlab 2015, only with Octave, it worked fine. 
I'll try 2015 later.

**Comment by nno** [comment]
I replicated this problem with Matlab 2017a (pre-release).
Although I have to do some more digging, at least the following seems to be causing issues in this recent version:

- in initTestSuite:

    ```
    if nargout(func)==0
    ```

  nargout gives an error that it cannot find the function.

- in initTestSuite:

    ```
    if ~nargout
    ```

  nargout gives an error that "You can only call nargin/nargout from within a MATLAB function."

So in summary, it seems that nargout has changed its behavior. I will try to find out more about this.

**Comment by homocomputeris** [comment]
Thanks! Then I won't test it on r2015.

**Comment by nno** [comment]
The issue seems introduces by Matlab 2016b, where the changes [1] read:

1) Compatibility Consideration: "Local functions are visible only to other functions in the same file. Previously, if a function file that defines local functions invoked a script, that script could call those local functions."

    Use this instead: "Store the function handle to the local function in a variable and use that instead of the call to the local function. For example, assign fcn = @myLocalFcn and update the script to use the variable fcn instead of calling myLocalFcn directly."


2) Compatibility Consideration: "If you call nargin, nargout, or inputname from within a script, MATLAB throws an error."

    Use this instead: "Move calls to nargin, nargout, and inputname into the relevant function file. If you need the results in a script, call the functions and store the results in a variable. Then use the variable in the script."

This is quite unfortunate, as these changes break how MOxUnit gets tests through ``initTestSuite``.
Indeed, the only approach I can see to get function handles to test functions is through the builtin ``localfunctions`` function, which unfortunately can only be used in the function itself, not in ``initTestSuite``. 

However, ``localfunctions`` does not exist in GNU Octave 4.0.3 (or earlier; it seems 4.1 will support it), so has to be surrounded with a try-catch statement (and we cannot use a helper function because then the local functions become outside the namespace). So, every test has to be rewritten from:

```
function test_suite=test_foo
    initTestSuite;
```

into

```
function test_suite=test_foo
    try % use of "localfunctions "is necessary in Matlab 2016b and later
        test_functions=localfunctions();
    catch % no problem; early Matlab versions can access local functions
    end
    initTestSuite;
```

and initTestSuite has to be modified to check for the presence of the ``test_functions`` variable. Ouch, this is far from pretty... but I don't see any other viable approach to keep compatibility with both recent Matlab and current (older) Octave versions. Any ideas @homocomputeris?

[1] https://it.mathworks.com/help/matlab/release-notes.html?rntext=&startrelease=R2015a&endrelease=R2016b&groupby=release&sortby=descending&searchHighlight=

**Comment by nno** [comment]
This should be addressed by #42, please re-open if not.
